[WIP] precise GC as presented at DConf 2013

rainers opened the pull request:
This is the precise GC as presented at DConf 2013. It generates pointer bitmaps for each type T using RTInfo!T and stores a bitmap alongside pool memory to use for following only pointer references.

RTInfo generation isn't working properly at the moment (needs https://github.com/D-Programming-Language/dmd/pull/3958 or https://github.com/D-Programming-Language/dmd/pull/2480), but using the compiler default will just scan some types conservatively instead of precisely.

Some notes:
- only the heap is scanned precisely, not stacks or the data/TLS segments
- compilation is slower: I measured 3:46min for the "quick" test suite on Win32 instead of 3:26 for master (we might need a special trait to avoid a lot of CTFE)
- execution is slower: the allocation heavy druntime benchmarks take 58sec instead of 51sec (unfortunately there are 2 list-benchmark  which don't give reliable timings).
- associative arrays need a dirty hack to support precise scanning of the key/value pairs in Entry: the TypeInfo for that is geneerated at runtime and stored in the Impl struct of each instance.
- dynamic arrays need to "re-emplace" their type info pointer bitmaps for large arrays, because data is "sometimes" placed at an offset of 16 bytes

I'm opening this as a pull request to verify building on other platforms aswell as making it more publically visible and open for review/discussion.

I plan to create another version that stores type info per allocation block and use this during scanning. This can avoid the troubles for the arrays, but will not allow reassigning new types to partial memory blocks (e.g. when using std.emplace on fields of a class).

rainers commented:
Updated to recent changes of the GC. With the getPointerBitmap trait, there is no longer a compile time penalty, but it's still 0-20% slower in the benchmarks, mostly during allocation. The benchmarks don't suffer much from false pointers, though.

MartinNowak commented on src/gc/gc.d:
```diff
@@ -2033,6 +2208,13 @@ struct Gcx
+    {
+        void* pbot;
+        void* ptop;
+        Pool* pool;
```
Twice as big and can't be returned in registers, that looks expensive.

MartinNowak replied:
Might be better to use a second stack for the pools in parallel.

rainers replied:
I was hoping for any function calls dealing with ScanRange to be inlined, though I didn't check yet.

MartinNowak replied:
Or you might try to pack the struct and change passing and return to Range + Pool*, using ref return for pop.

rainers replied:
I checked the disassembly: it's all inlined AFAICT. So the struct needs 50% more memory (there is no padding), but there is probably not a lot we can do here. The non-precise version might use the Range struct though, so we could templatize ToScanStack on that type.
Indexing might be a little more expensive than with a struct size of 8/16, so we could also use a pointer to walk the stack.

MartinNowak commented on src/gc/gc.d:
```diff
@@ -2910,6 +3117,8 @@ struct Pool
 
         mark.alloc(nbits);
+        if(GC.config.precise)
+            is_pointer.alloc(cast(size_t)poolsize/(void*).sizeof);
```
Might want to allocate those lazily.

rainers replied:
Ok, might make sense for large object pools.

rainers replied:
I just tried lazy allocation of Pool.is_pointer, but it makes most of the benchmarks run a little slower. I don't think there are many pools without pointers.

MartinNowak commented on src/object.di:
```diff
@@ -679,11 +679,18 @@ bool _xopCmp(in void* ptr, in void* ptr);
 template RTInfo(T)
 {
-    enum RTInfo = cast(void*)0x12345678;
+    enum RTInfo = __traits(getPointerBitmap,T).ptr;
```
If you pass the bitmap array as value argument through a template, they'll get deduplicated.

rainers replied:
Ok, I'll try, though I doubt it will be a significant space reduction.

rainers replied:
Ouch, the above actually allocates every time RTInfo!T is used in TypeInfo.rtInfo().

MartinNowak commented:
I'll have a look at the benchmarks, we definitely need to get this fast enough.

MartinNowak commented on src/gc/gc.d:
```diff
@@ -2122,6 +2308,12 @@ struct Gcx
                 {
+                    static if (precise) if (p1pool)
+                    {
+                        size_t p1off = cast(size_t)(p1 - cast(void**)p1pool.baseAddr);
```
You should move this test to before p1 is dereferenced. If something isn't a pointer we should perform as little work as possible. This loop is key to make this thing fast.

rainers replied:
The previous versions had some optimization to this respect, but I wanted to make minimal changes to begin with. The benchmarks don't show a large increase in scanning time, the preformance decrease is caused by setting the pointer bitmaps.

MartinNowak replied:
But it might be possible to make up for this by improving the marking time.

rainers replied:
I just tried it: without further optimizations, moving the code above `p = *p1` slows down marking by up to 10%.

rainers commented on src/rt/aaA.d:
```diff
@@ -178,7 +236,7 @@ body
     if (aa.impl is null)
     {
-        aa.impl = new Impl();
+        aa.impl = new Impl(keyti, typeid(void*), valuesize);
```
What's going on with the templated AA implementation? Would it hurt to add the value type information with #1037 and https://github.com/D-Programming-Language/dmd/pull/4159?

MartinNowak replied:
Work on AA is frozen until we have a better implementation.

MartinNowak commented on src/object.di:
```diff
@@ -679,11 +679,23 @@ bool _xopCmp(in void* ptr, in void* ptr);
+template RTInfoImpl(T)
+{
+    immutable RTInfoImpl = __traits(getPointerBitmap,T);
+}
```
This would actually get you the deduplication.

```d
template RTInfoImpl(ubyte[] pointers)
{
    immutable data = pointers;
    immutable RTInfoImpl = &data;
}

template RTInfo(T)
{
    immutable RTInfo = RTInfoImpl!(__traits(getPointerBitmap, T));
}
```

rainers commented on src/rt/aaA.d:
```diff
@@ -480,7 +545,7 @@ inout(ArrayRet_t) _aaKeys(inout AA aa, in size_t keysize, const TypeInfo tiKeyAr
         return null;
 
-    void[] res = _d_newarrayU(tiKeyArray, len);
+    void* res = _d_newarrayU(tiKeyArray, len).ptr;
```
This change is unrelated, but the current version causes a range error when building druntime without -release.
The return value of _d_newarrayU is confusing, as it actually returns the length for `T[]`, not `void[]`. It should rather just return `void*`

MartinNowak commented:
Delay until we have an idea how to get this fast, e.g. 2.068?

rainers commented:
> Delay until we have an idea how to get this fast, e.g. 2.068?

Yes. I'm also waiting for either https://github.com/D-Programming-Language/dmd/pull/2480 or https://github.com/D-Programming-Language/dmd/pull/3958 making some progress. This GC implementation still works with the incomplete RTInfo by making conservative assumptions, though.

rainers commented:
Rebased, added small optimization and fixed a crash with the vdparser benchmark.

The precise GC is now a little faster than the standard GC for the vdparser, but still notably slower on some others that don't benefit from preciseness.

rainers commented:
Fixed test and rebased.

MartinNowak commented:
Hope I'll find time to help you getting this fast.
Have you played around with compaction already? This could be good for a few percent speedup.
You just need to look objects that are reference by untyped/unprecise objects.

rainers commented:
> Hope I'll find time to help you getting this fast.

I recently tried to figure out what is actually causing the overhead. It seems roughly equally distributed between the three functional parts of `setPointerBitmap`:
- calling the virtual function `rtInfo`
- figuring out what bits to copy
- actually copy the appropriate bits

But I'm currently out of ideas how to get these substantially faster. It's also no fun to optimize against the dmd backend...

The overhead during scanning is small compared to the allocation time, but we could try to take advantage of not scanning some pointers. The possible gain is not large, though, as long as we actually don't detect false pointers.

rainers commented:
> Have you played around with compaction already? This could be good for a few percent speedup.
> You just need to pin objects that are reference by untyped/unprecise objects.

Not yet. You will also have to add another bit to the type info to distinguish 'exact' pointers from possible false pointers, e.g. inside unions. I'm not sure we gain much by moving objects around as long as they are kept in free lists, so that would have to be changed, too.

rainers commented:
BTW: as shown in the graph https://gist.github.com/rainers/6cdf73b48837defb9f88, people lived with a slower version of the GC before 2.067, so they might accept a small performance step backwards for some application in exchange for a precise GC.

MartinNowak commented:
Well being precise isn't a value in itself, smaller memory footprint and faster GC are.

MartinNowak commented:
From everything that I've, except for a few of our benchmarks that fit in L3, the biggest problem is memory bandwidth/latency. There are quite some cycles to spend if you can only avoid fetching a cacheline. If OTOH you're putting more pressure on the cache/mem by adding more metadata it's really hard to win.
Maybe we can use something smarter than a bit vector, which comes with a 1/64=1.6% metadata overhead.

rainers commented:
I had to reset this one as rebase continued to produce merge conflicts. Some of the intermediate commits were obsolete anyway.

rainers commented:
> Well being precise isn't a value in itself, smaller memory footprint and faster GC are.

The benchmarks are mostly using too little memory for a short time only, so preciseness does not show its usefullness there. The vdparser test shows a smaller memory footprint for 32-bit (sometimes also on 64-bit) with dramatically smaller pause times. Raising the number of iterations easily makes the precise version run faster, with the normal version running out of memory at some point.

rainers commented:
> Maybe we can use something smarter than a bit vector, which comes with a 1/64=1.6% metadata overhead.

Do you have an idea how we can reduce this overhead? Saving type information per object proved to be a lot slower due to the additional indirections (I tried that in https://github.com/D-Programming-Language/druntime/pull/1057)

MartinNowak commented:
> Raising the number of iterations easily makes the precise version run faster, with the normal version running out of memory at some point.

We should definitely add a leak test to our benchmark suite.

MartinNowak commented:
> Do you have an idea how we can reduce this overhead? Saving type information per object proved to be a lot slower due to the additional indirections (I tried that in #1057)

I hope I can get back onto the GC stuff next week. So far I don't have a simple idea/solution, just the notion, that copying around lots of bits for each object is slow.
To improve on this one would need to deduplicate identical metadata, e.g. by grouping objects with the same types/bitpatterns.

rainers commented:
Rebased.

drbo commented:
What is the status of this?  Will this be merged?

rainers commented:
Rebased in preparation of the PR's first anniversary.

ZombineDev commented:
@rainers what's the status of this PR? Are there any serious showstoppers before this can be merged?

rainers commented:
> Are there any serious showstoppers before this can be merged, or is it only the performance that you want to improve?

AFAICT this is good enough to be included. @MartinNowak has some reservations regarding performance, but I don't see a faster solution in the near future. I tried some microtuning in the past, but that is rather frustrating when done against the dmd backend.

Before actual merging I'd make it opt-in rather than opt-out. Presice scanning is enabled in this PR so the auto-tester is actually running any of the new code.

What's missing is some documentation (some notes on what has to be done to get your manually managed memory scanned precisely), but I'd rather postpone writing it until this PR has a good chance of being merged.

rainers commented:
Rebased. Needs https://github.com/D-Programming-Language/dmd/pull/5566 to build phobos unittests, though.

drbo commented:
I would love to see this.  Next step is incremental concurrent collection.

MartinNowak commented:
> Next step is incremental concurrent collection.

It is unlikely that we'll add write barriers, so we can't use classical techniques (e.g. generational) for incremental collection.

rainers commented:
Closing in favor of https://github.com/dlang/druntime/pull/1603
